### DISCUSSION TITLE
开发鸿蒙 && OpenHarmony API 库

### DISCUSSION THREAD

**Discussion by Chen-jj** [NervJS/taro#9645] [鸿蒙 && OpenHarmony 适配小组]
使用鸿蒙的 API 去封装遵循微信小程序规范的 API。

### 1. 环境准备

首先按照 [《试用编译鸿蒙应用的功能》](https://github.com/NervJS/taro/discussions/9641) 中的步骤，准备好调试环境。

### 2. 开发

以开发微信小程序规范的**路由功能** API 为例。

#### 2.1 创建目录与文件

进入 `taro-harmonyos` 包的 `src` 目录，在 `src/apis` 中创建 `router/index.ts`。

#### 2.2 开发

> 引用鸿蒙原生 API 时，请使用 `require` 的方式，
> 如：`const router = require('@system.router')`

编辑 `src/apis/router/index.ts` 文件进行开发，导出需要的 API。

#### 2.3 导出

编辑 `src/apis/apis.ts` 文件，使用 `export from` 语法导出 `src/apis/router/index.ts` 的 API。

### 3. 调试

在 Taro 项目中使用相应的 API 进行调试，如：`Taro.navigateTo(...)`

## 规划

需要优先支持的 API：

### 用户

- login 登录
- getUserInfo 获取用户信息
- getUserProfile 获取用户信息
- authorize 提前向用户发起授权请求  鸿蒙不支持
- getSetting 获取用户的当前设置  鸿蒙不支持
- openSetting 调起客户端小程序设置界面，返回用户设置的操作结果  鸿蒙不支持

### 存储

- setStorage
- setStorageSync
- getStorage
- getStorageInfo
- getStorageInfoSync
- getStorageSync
- removeStorage
- removeStorageSync
- clearStorage
- clearStorageSync

### 系统

- getNetworkType ： 已实现 -  remote device模拟器测试通过
- onNetworkStatusChange ： 已实现 -  remote device模拟器测试通过
- getSystemInfo ： 已实现 -  remote device模拟器测试通过  
          brand: 已实现//设备品牌
          model:已实现//设备型号
          pixelRatio: 不支持//设备像素比	
          screenWidth: 已实现//屏幕宽度，单位px  需api7支持
          screenHeight: 已实现//屏幕高度，单位px  需api7支持
          windowWidth: 已实现//可使用窗口宽度，单位px
          windowHeight: 已实现//可使用窗口高度，单位px
          statusBarHeight: 不支持//状态栏的高度，单位px
          language:已实现   需api7支持
          version: 已实现//版本号
          system: 已实现//操作系统及版本
          platform: 'android',//客户端平台
          fontSizeSetting: 不支持//用户字体大小（单位px）
          SDKVersion: 已实现//客户端基础库版本
          benchmarkLevel: 不支持//设备性能等级
          albumAuthorized: 不支持//允许微信使用相册的开关（仅 iOS 有效）
          cameraAuthorized:不支持//允许微信使用摄像头的开关
          locationAuthorized: 不支持//定位的开关
          microphoneAuthorized: 不支持//麦克风的开关
          notificationAuthorized: 不支持//通知的开关
          notificationAlertAuthorized: 不支持//通知带有提醒的开关（仅 iOS 有效）
          notificationBadgeAuthorized:不支持//通知带有标记的开关（仅 iOS 有效）
          notificationSoundAuthorized: 不支持//通知带有声音的开关（仅 iOS 有效）
          phoneCalendarAuthorized:不支持//使用日历的开关
          bluetoothEnabled: 不支持//蓝牙的系统开关
          locationEnabled:不支持//地理位置的系统开关
          wifiEnabled: 不支持//Wi-Fi 的系统开关
          safeArea:不支持//在竖屏正方向下的安全区域
          locationReducedAccuracy: 不支持//模糊定位精准定位
          theme:不支持//系统当前主题，取值为light或dark
          host: 不支持//当前小程序运行的宿主环境	
          enableDebug: 不支持//是否已打开调试
          deviceOrientation:已实现//设备方向 api7支持
- getSystemInfoSync ： 已实现 -  remote device模拟器测试通过
- getScreenBrightness ： 已实现 -  remote device模拟器测试通过
- setScreenBrightness ： 已实现 -  remote device模拟器测试通过
- onMemoryWarning  不支持
- offMemoryWarning  不支持
- makePhoneCall ： 已实现  - remote device模拟器上测试不行，需要真机测试 需api7支持
- setClipboardData ： 已实现 -  remote device模拟器测试通过
- getClipboardData ： 已实现 -  remote device模拟器测试通过

### 网络

- request 已实现
- RequestTask.abort 已实现
- RequestTask.offHeadersReceived 已实现
- RequestTask.onHeadersReceived 已实现
- downloadFile 已实现 因使用了on('complete'),on('fail'),query等API7支持的接口，所以API7支持
- DownloadTask.abort 已实现
- DownloadTask.offProgressUpdate 已实现
- DownloadTask.onProgressUpdate 已实现
- uploadFile 已实现
- UploadTask.abort 已实现
- UploadTask.offHeadersReceived 已实现 API7支持
- UploadTask.offProgressUpdate 已实现
- UploadTask.onHeadersReceived 已实现 API7支持
- UploadTask.onProgressUpdate 已实现
- connectSocket 已实现
- SocketTask.close 已实现
- SocketTask.onClose 已实现
- SocketTask.onError 已实现
- SocketTask.onMessage 已实现
- SocketTask.onOpen 已实现
- SocketTask.send 已实现
- closeSocket 不支持
- onSocketOpen 不支持
- onSocketMessage 不支持
- onSocketError 不支持
- onSocketClose 不支持
- sendSocketMessage 不支持
- DownloadTask.offHeadersReceived 此接口 ohos 不支持
- DownloadTask.onHeadersReceived 此接口 ohos 不支持

### 交互

- showModal
- showLoading
- hideLoading
- showToast
- hideToast
- showActionSheet
- showShareMenu
- hideShareMenu

### 路由

- navigateTo
- navigateBack
- redirectTo

### 界面

- pageScrollTo (SDK API Version 6才支持，需安装EevEco Studio 3.0 Beta)

### 位置 **从API Version 7 开始支持**

- getLocation：已实现
- stopLocationUpdate：不支持
- startLocationUpdate：不支持
- startLocationUpdateBackground：不支持
- onLocationChangeError：不支持
- onLocationChange：已实现
- offLocationChangeError：不支持
- offLocationChange：已实现
- chooseLocation：无法实现，与腾讯地图相关，OpenHarmony暂不支持
- openLocation：无法实现，与腾讯地图相关，OpenHarmony暂不支持
- choosePoi：无法实现，与腾讯地图相关，OpenHarmony暂不支持

### 节点

- createSelectorQuery
- createIntersectionObserver

### 媒体

- getImageInfo 已实现 API7
- saveImageToPhotosAlbum 已实现
- previewImage 已实现
- chooseImage 已实现
- chooseMessageFile 不支持
- createVideoContext 不支持，需额外封装组件

### 画布

- createCanvasContext

### 动画

- createAnimation 
    - 不支持 transformOrigin 参数
    - 返回的 Animation 对象，不支持 export、step、matrix、matrix3d、rotate、rotate3d 等等一系列微信小程序支持的方法。

### 支付

- requestPayment

### 下拉刷新

- startPullDownRefresh
- stopPullDownRefresh
  - 容器 `overflow: scroll` 时不能下拉
  - 调用 `startPullDownRefresh` 时没有动画，手动下拉时有动画

### 其它

- nextTick
- canIUse

### 鸿蒙不支持

- switchTab
- showTabBar
- hideTabBar
- setNavigationBarTitle
- setNavigationBarColor
- showNavigationBarLoading
- hideNavigationBarLoading
- getMenuButtonBoundingClientRect
- hideHomeButton
- chooseLocation
- openLocation
- choosePoi
- DownloadTask.offHeadersReceived
- DownloadTask.onHeadersReceived
- closeSocket
- onSocketOpen
- onSocketMessage
- onSocketError
- onSocketClose
- sendSocketMessage

**Comment by chuckchen314** [comment]
### **适配鸿蒙原生API**

**背景**
鸿蒙的API比小程序的不管是种类还是数量都很多，比如跨端迁移和多端协同，完全小程序应该是不可能的。如果逐个的适配鸿蒙API，也是不可行的。

**实现方法**

1. 将鸿蒙SDK中的JS下的接口定义(typescript definition)打包发布至npm
2. 扩展taro中webpack的externals，使鸿蒙接口调用导入路径正常。如下：
`import File from '@harmony/api/phone/@system.file'`
`taro webpack编译后：module.exports = require('@system.file')`

**结果**
验证OK

**Reply by Chen-jj** [reply]
是的，Taro 只会尽量把小程序规范里的 API 使用鸿蒙 API 进行实现。而鸿蒙 API 甚至其它平台的 API 肯定是小程序规范的超集，Taro 会做到支持调用平台的 API，对于鸿蒙来说就是你提到的方法。
